src/repository/opamRepository.ml

```diff
@@ -245,20 +245,66 @@ let pull_from_mirrors label ?working_dir ?subpath cache_dir destdir checksums ur
     url, Not_available (Some m, m)
   | ret -> ret
 
-let pull_tree
-    label ?cache_dir ?(cache_urls=[]) ?working_dir ?subpath
-    local_dirname checksums remote_urls =
-  let extract_archive f s =
-    OpamFilename.cleandir local_dirname;
-    OpamFilename.extract_job f local_dirname @@+ function
-    | None -> Done (Up_to_date s)
-    | Some (Failure s) ->
-      Done (Not_available (Some s, "Could not extract archive:\n"^s))
-    | Some (OpamSystem.Process_error pe) ->
-      Done (Not_available (Some (OpamProcess.result_summary pe),
-                           OpamProcess.string_of_result pe))
-    | Some e -> Done (Not_available (None, Printexc.to_string e))
+(* handle subpathes *)
+let pulls_tree_t
+    ?cache_dir ?(cache_urls=[]) ?working_dir ?subpath
+    dirnames checksums remote_urls =
+  let extract_archive =
+    let fallback success = function
+      | None -> success ()
+      | Some (Failure s) ->
+        Done (Not_available (Some s, "Could not extract archive:\n"^s))
+      | Some (OpamSystem.Process_error pe) ->
+        Done (Not_available (Some (OpamProcess.result_summary pe),
+                             OpamProcess.string_of_result pe))
+      | Some e -> Done (Not_available (None, Printexc.to_string e))
+    in
+    match dirnames with
+    | [ _label, local_dirname ] ->
+      (fun archive msg ->
+         OpamFilename.cleandir local_dirname;
+         OpamFilename.extract_job archive local_dirname
+         @@+ fallback (fun () ->  Done (Up_to_date msg)))
+    | _ ->
+      fun archive msg ->
+        OpamFilename.with_tmp_dir_job @@ fun tmpdir ->
+        let copies () =
+          OpamParallel.map ~jobs:3
+            ~command:(fun (label, local_dirname) ->
+                let text = OpamProcess.make_command_text label label in
+                OpamProcess.Job.with_text text @@
+                (try
+                   OpamFilename.cleandir local_dirname;
+                   OpamFilename.copy_dir ~src:tmpdir ~dst:local_dirname;
```

Comment:
I am wondering if, in terms of performance, it may not actually be better to extract n times rather than extract once and copy… In any case extracting directly to one of the `dirnames` rather than a tmpdir could remove a step, and enable to discard the special case for a single target dir ?

Reply:
I kept the same old behavior of : extract in a temp directory then copy to dirname. See discussion in #5018

Reply:
ha! Indeed; so if we want to optimise we'll need to both extract many times and optimise extraction to avoid the copy. Does Kate's PR already do this ? an option could be to have the tmp dir besides the target dir (to ensure they're on the same fs) then move instead of copy…